src/v/storage/log_manager.cc

```diff
@@ -127,31 +130,49 @@ ss::future<> log_manager::housekeeping() {
      * them as fast as we can since the majority of the time they will be in the
      * hotpath / (request-response)
      */
+    std::unique_ptr<model::ntp> current_key{};
     using bflags = log_housekeeping_meta::bitflags;
-    return ss::do_until(
-             [this] {
-                 auto it = find_next_non_compacted_log(_logs);
-                 return it == _logs.end();
-             },
-             [this, collection_threshold] {
-                 return ss::with_scheduling_group(
-                   _config.compaction_sg, [this, collection_threshold] {
-                       auto it = find_next_non_compacted_log(_logs);
-                       if (it == _logs.end()) {
-                           // must check again because between the stop
-                           // condition and this continuation we might have
-                           // removed the log
-                           return ss::now();
-                       }
-                       it->second.flags |= bflags::compacted;
-                       it->second.last_compaction = ss::lowres_clock::now();
-                       return it->second.handle.compact(compaction_config(
+    return ss::repeat([this,
+                       collection_threshold,
+                       current_key = std::move(current_key)]() mutable {
+               return ss::with_scheduling_group(
+                 _config.compaction_sg,
+                 [this,
+                  collection_threshold,
+                  current_key = std::move(current_key)]() mutable {
+                     typename logs_type::iterator it, start_it;
+
+                     if (!current_key) {
+                         current_key.reset(new model::ntp());
+                         start_it = _logs.begin();
+                     } else {
+                         start_it = _logs.find(*current_key);
+                     }
+
+                     if (start_it != _logs.end()) {
+                         it = find_next_non_compacted_log(_logs, start_it);
+                     } else {
+                         // The ntp was deleted.
+                         it = find_next_non_compacted_log(_logs, _logs.begin());
```

Comment:
starting at the beginning when the current_key isn't found seems completely reasonable. i am wondering though if it would make sense to use something like `_logs.lower_bound` or `_logs.upper_bound` such that if current_key were to be removed, then the search would start at whatever would have come before current_key. so it might result in some duplicates, but not necessarily from the very beginning. again just thinking about this; there may be some cases i'm not thinking about.